explorer/src/commitUtils.js

```diff
@@ -0,0 +1,93 @@
+// @flow
+
+type CommitData = {
+  // TODO improve variable names
+  fileToCommits: {[filename: string]: string[]};
+  commits: {[commithash: string]: Commit};
+  authors: string[];
+}
+
+type Commit = {
+  author: string;
+  stats: {[filename: string]: FileStats};
+}
+
+type FileStats = {
+  lines: number;
+  added: number;
+  deleted: number;
+}
+
+export function commitWeight(commit: Commit, filepath: string): number {
+  // hack - GitPython encodes renames in the filepath. ignore for now.
+  if (filepath.indexOf('=>') !== -1) {
+    return 0;
+  }
+  return Math.sqrt(commit.stats[filepath].lines);
+}
+
+function allSelectedFiles(filepath: string, data: CommitData): string[] {
+  const fnames = Object.keys(data.fileToCommits);
+  return fnames.filter(x => x.startsWith(filepath));
+}
+
+function *userWeights(files: string[], data: CommitData, weightFn: WeightFn): Iterable<[string, number]> {
+  for (const file of files) {
+    for (const commitHash of data.fileToCommits[file]) {
+      const commit = data.commits[commitHash];
+      let w;
+      if (commit.stats[file] == null) {
+        // hack - likely due to the GitPython file rename issue
+        console.log(`commit ${commitHash} missing file ${file}`);
+        w = 0;
+      } else {
+        w = weightFn(commit, file);
+      }
+      yield [commit.author, w];
+    }
+  }
+}
+
+type WeightFn = (commit: Commit, filepath: string) => number;
```

Comment:
I think this merits a docstring, to make it clear that weights are aggregated for a file, and that this does not compute the "weight of a commit". What do you think of something like

> A weight function determines how much commit contributes to the importance of a particular file. For instance, a weight function might be defined as, “the number of lines changed in `filepath` in commit `commit`”, or “`1` if `filepath` was changed in `commit`, else `0`”.

?

Reply:
Done